std/mem.zig

```diff
@@ -84,6 +84,36 @@ pub const Allocator = struct {
     }
 };
 
+pub var c_allocator = Allocator {
+    .allocFn = cAlloc,
+    .reallocFn = cRealloc,
+    .freeFn = cFree,
+};
+
+const c = @cImport(@cInclude("stdlib.h"));
+
+fn cAlloc(self: &Allocator, n: usize, alignment: usize) -> %[]u8 {
+    if (c.malloc(usize(n))) |mem| {
+        @ptrCast(&u8, mem)[0..n]
+    } else {
+        error.OutOfMemory
+    }
+}
+
+fn cRealloc(self: &Allocator, old_mem: []u8, new_size: usize, alignment: usize) -> %[]u8 {
+    const old_ptr = @ptrCast(&c_void, &old_mem[0]);
+    if (c.realloc(old_ptr, usize(new_size))) |mem| {
```

Comment:
Is libc `realloc` guaranteed to succeed if `new_size <= old_mem.len`? If yes, this code is correct. If no, we should handle failure in the case that `new_size <= old_mem.len` by returning the same slice with a new len, to conform to the zig allocator semantics for realloc.

Reply:
C99 standard seems to state that memory is always allocated and moved even on smaller sizes. Added this logic.